src/split_at.rs

```diff
@@ -233,59 +213,571 @@ unsafe impl<T> SplitAt for [T] {
+    ///     // return concurrent references to `packet` and `rest`.
+    ///     let (packet, rest) = split.via_into_bytes();
+    ///
+    ///     assert_eq!(packet.length, 4);
+    ///     assert_eq!(packet.body, [1, 2, 3, 4]);
+    ///     assert_eq!(rest, [5, 6, 7, 8, 9]);
+    ///
+    ///     rest.fill(0);
+    /// }
+    ///
+    /// assert_eq!(packet.length, 4);
+    /// assert_eq!(packet.body, [1, 2, 3, 4, 0, 0, 0, 0, 0]);
+    /// ```
+    #[must_use = "has no side effects"]
+    #[inline(always)]
+    pub fn via_into_bytes(self) -> (&'a mut T, &'a mut [T::Elem])
+    where
+        T: IntoBytes,
+    {
+        let (l, r) = self.into_ptr().via_into_bytes();
+        (l.as_mut(), r.as_mut())
+    }
+
+    /// Produces the split parts of `self`, using [`Unaligned`] to ensure that
+    /// it is sound to have concurrent references to both parts.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use zerocopy::{SplitAt, FromBytes};
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(SplitAt, FromBytes, KnownLayout, IntoBytes, Unaligned)]
+    /// #[repr(C)]
+    /// struct Packet<B: ?Sized> {
+    ///     length: u8,
+    ///     body: B,
+    /// }
+    ///
+    /// // These bytes encode a `Packet`.
+    /// let mut bytes = &mut [4, 1, 2, 3, 4, 5, 6, 7, 8, 9][..];
+    ///
+    /// let packet = Packet::<[u8]>::mut_from_bytes(bytes).unwrap();
+    ///
+    /// assert_eq!(packet.length, 4);
+    /// assert_eq!(packet.body, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
+    ///
+    /// {
+    ///     // Attempt to split `packet` at `length`.
+    ///     let split = packet.split_at_mut(packet.length as usize).unwrap();
+    ///
+    ///     // Use the `Unaligned` bound on `Packet` to prove that it's okay to
+    ///     // return concurrent references to `packet` and `rest`.
+    ///     let (packet, rest) = split.via_unaligned();
+    ///
+    ///     assert_eq!(packet.length, 4);
+    ///     assert_eq!(packet.body, [1, 2, 3, 4]);
+    ///     assert_eq!(rest, [5, 6, 7, 8, 9]);
+    ///
+    ///     rest.fill(0);
+    /// }
+    ///
+    /// assert_eq!(packet.length, 4);
+    /// assert_eq!(packet.body, [1, 2, 3, 4, 0, 0, 0, 0, 0]);
+    /// ```
+    #[must_use = "has no side effects"]
+    #[inline(always)]
+    pub fn via_unaligned(self) -> (&'a mut T, &'a mut [T::Elem])
+    where
+        T: Unaligned,
+    {
+        let (l, r) = self.into_ptr().via_unaligned();
+        (l.as_mut(), r.as_mut())
+    }
+
+    /// Produces the split parts of `self`, using a dynamic check to ensure that
+    /// it is sound to have concurrent references to both parts. You should
+    /// prefer using [`Self::via_into_bytes`] or [`Self::via_unaligned`], which
+    /// have no runtime cost.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use zerocopy::{SplitAt, FromBytes};
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(SplitAt, FromBytes, KnownLayout, IntoBytes)]
```

Comment:
Presumably you don't want to derive `IntoBytes`? As a consequence, you presumably don't need `Packet` to be generic over `B`?

Reply:
It's required by `FromBytes::mut_from_bytes`.

Reply:
Ack. I wonder if this is a hint that users will ~never need this API lol. In particular, since slice DSTs are unsized, you can't ever construct one by value, and so the only way to construct a `&mut MySliceDst` is via reference transmute, but all of our support for reference transmutes requires `Self: IntoBytes`. I suppose you could have constructed one yourself (e.g. via unsized coercion), but it's certainly niche. Oh well.

Reply:
A bunch of these look a little silly when situated right next to `FromByte::ref_from_bytes` (which requires `Immutable`), and `FromBytes::mut_from_bytes` (which requires `IntoBytes`), but I can see these being useful in generic contexts.

Reply:
Example?

Reply:
Let's say you have a generic function that needs to split a dst that gets passed in. This function is generic over `&` and `&mut` references to a DST. Maybe you have some `&mut` DSTs that aren't `Immutable`, and some `&` DSTs that aren't `IntoBytes`. In such a case, a `T: Unaligned` bound might be the common denominator that lets you write this generic function.

Reply:
I see `SplitAt` as extending the toolkit of DST manipulation that's dependent on zerocopy traits and expertise to get right, towards a goal of working with them nearly as easily as sized structs. So I do generally support whatever flexible trait bounds are possible for converting, even though I don't have a "key use case" example either.

> you can't ever construct one by value

I've been thinking that `Box<Dst>` construction could be done with a macro and struct init syntax. This'd fit well in zerocopy with the knowledge from `KnownLayout`.

Reply:
> In such a case, a `T: Unaligned` bound might be the common denominator that lets you write this generic function.

Sure, but this comment thread is about `via_runtime_check`, not about `via_unaligned`. My point is that, if you *always* need to have an `IntoBytes` bound anyway, then you may never actually need `via_runtime_check`.

---

To be clear, I still think we should keep `via_runtime_check` for completeness. I just think it's an interesting observation that we should think more about at some point.

Reply:
> I've been thinking that `Box<Dst>` construction could be done with a macro and struct init syntax. This'd fit well in zerocopy with the knowledge from `KnownLayout`/`SplitAt`.

Oh this could be really cool! If you have a prototype at some point, let us know! I could see adding this to zerocopy if it's a simple enough API (which I imagine it would be).